storage/database/db_manager.go

```diff
@@ -42,6 +42,9 @@ var (
 	logger = log.NewModuleLogger(log.StorageDatabase)
 
 	errGovIdxAlreadyExist = errors.New("a governance idx of the more recent or the same block exist")
+
+	backupHashes [128]common.Hash
```

Comment:
In the PR description, backup stores the hash before 64 blocks, and yet this size is 128.
Doesn't this mean that when storing hash of block N, backup hash stores block N-128?

Also, is block N-128 guaranteed (or probably guaranteed) to have block contents? I wonder if the number has a rationale behind.

Reply:
My initial intention was 128 blocks as ian said. So I made it into an array of 128. However, since there are two places to store the headhash (WriteHeadBlockHash, WriteHeadFastBlockHash), the result is N-64 blocks

Reply:
Oh, I see. I think it's quite hard to understand for readers.
What do you think about introducing a new object for backup hash?
I see it as a variant of _queue_ with unique IDs that can be only popped when it is full.
I think it's more readable this way.

```go
const backupHashCnt 64

struct backupHash {
    backupHashes [backupHashCnt]common.Hash
    idx int
}

func (b *backupHash) push(h common.Hash) { /* skip if duplicate; push to backupHashes */ }
func (b *backupHash) pop() common.Hash { /* if not full, return empty hash; otherwise, return N-64 */ }
```

Reply:
Yes. Changes to queue structure